### PR TITLE
initial commit

### PR DESCRIPTION
**Pull request by BroderickCarlin**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+[package]
+name = "pwinty"
+version = "0.1.0"
+authors = ["Broderick Carlin <broderick.carlin@gmail.com>"]
+edition = "2018"
```
Comment (TDHolmes) [inline Cargo.toml:5]: License! :D

File: src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        order_info: &payloads::OrderCreate,
+    ) -> Box<dyn Future<Item = payloads::Order, Error = ApiError> + Send> {
+        let body = match serde_json::to_string(&order_info) {
+            Ok(body) => body,
+            Err(e) => return Box::new(future::err(ApiError::InternalError(format!("{:?}", e)))),
+        };
+
+        Box::new(
+            self.make_request::<payloads::OrderWrapper>("/orders", Method::POST, true, Some(body))
+                .and_then(|order| future::ok(order.data)),
```
Comment (BroderickCarlin) [inline src/lib.rs]: Thoughts on `future::ok()` vs `Ok()`?

File: Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+[dependencies]
+reqwest = "0.9"
+serde = "1.0"
+serde_derive = "1.0"
+serde_json = "1.0"
+futures = "0.1"
+tokio = { version = "0.1.7"}
+http = "0.1"
+chrono = { version = "0.4", features = ["serde"] }
```
Comment (TDHolmes) [inline Cargo.toml]: nit: newline

File: src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    }
+}
+
+// impl<T: DeserializeOwned> Future for ApiError {
+//     type Item = T;
+//     type Error = ApiError;
+//     fn poll(&mut self) -> Poll<T, Self::Error> {
+//         Err(self)
+//     }
+// }
```
Comment (TDHolmes) [inline src/lib.rs]: booooo

File: src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    fn url_for_endpoint(&self, endpoint: &str) -> String {
+        let mut url: String = self.base_url.to_string();
+        url.push_str(self.version);
+        url.push_str(endpoint);
+        url
+    }
+
+    fn add_headers(&self, builder: RequestBuilder) -> Result<RequestBuilder, ApiError> {
+        let mut headers = HeaderMap::new();
+        headers.insert("X-Pwinty-MerchantId", self.merchant_id.parse()?);
```
Comment (GeorgeHahn) [inline src/lib.rs]: Consider parsing these on API construction so you can catch errors earlier

File: src/lib.rs (hunk 4 above)
Comment (BroderickCarlin) [reply src/lib.rs]: :nuked:

File: src/lib.rs (hunk 2 above)
Comment (TDHolmes) [reply src/lib.rs]: I don't know tokio / async await well enough to say one way or the other haha

File: src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        headers.insert("X-Pwinty-REST-API-Key", self.api_key.parse()?);
+        headers.insert("Content-Type", "application/json".parse()?);
+        headers.insert("accept", "application/json".parse()?);
```
Comment (GeorgeHahn) [inline src/lib.rs]: It looks like the idiomatic function to create headers from static strings is `HeaderValue::from_static`

File: src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    fn make_request<Resp: 'static + DeserializeOwned + Send>(
+        &self,
+        endpoint: &str,
+        method: Method,
+        add_headers: bool,
+        body: Option<String>,
+    ) -> Box<dyn Future<Item = Resp, Error = ApiError> + Send> {
+        let url = self.url_for_endpoint(endpoint);
+        let client = if add_headers {
+            match self.add_headers(Client::new().request(method, &url)) {
```
Comment (GeorgeHahn) [inline src/lib.rs]: Is it common to create a new client for each reqwest request? Just checking - some libraries require this, others prefer avoiding it.

**Review by GeorgeHahn — COMMENTED** [review]
I think the ergonomics of creating an order and adding images could be improved with a builder struct. Otherwise this looks great!

File: src/lib.rs (hunk 7 above)
Comment (BroderickCarlin) [reply src/lib.rs]: Looking back at the reqwests crate it does appear that they fall more into the mindset of using a shared Client. Will see if I can refactor a bit to go that route

File: src/lib.rs (hunk 6 above)
Comment (BroderickCarlin) [reply src/lib.rs]: Ohhhh, that should allow this to be rewritten to completely remove the `Result<>`

**Pull request merged**